zndraw/frame.py

```diff
@@ -0,0 +1,191 @@
+
+        try:
+            calc_data = {}
+            for key in atoms.calc.results:
+                value = atoms.calc.results[key]
+                if isinstance(value, np.ndarray):
+                    value = value.tolist()
+                calc_data[key] = value
+
+            frame.calc = calc_data
+        except (RuntimeError, AttributeError):
+            pass
+
+        return frame
+
+    def to_atoms(self) -> ase.Atoms:
+        atoms = ase.Atoms(
+            positions=self.positions,
+            numbers=self.numbers,
+            cell=self.cell,  # self.cell funktioniert wieso auch immer nicht. TODO
+            pbc=self.pbc,
+        )
+
+        # atoms.arrays["colors"] = self.colors
+
+        atoms.connectivity = nx.Graph()
+        for edge in self.connectivity:
+            atoms.connectivity.add_edge(edge[0], edge[1], weight=edge[2])
+
+        if self.calc is not None:
+            atoms.calc = SinglePointCalculator(atoms)
+            atoms.calc.results = {
+                key: np.array(val) if isinstance(val, list) else val
+                for key, val in self.calc.items()
+            }
+
+        return atoms
+
+    def calc_bonds(self):
+        """
+        Experimental Function, currently not in use
+        """
+        single_bond_multiplier: float = Field(1.2, le=2, ge=0)
+        double_bond_multiplier: float = Field(0.9, le=1, ge=0)
+        triple_bond_multiplier: float = Field(0.0, le=1, ge=0)
+
+        cutoffs = [
+            single_bond_multiplier,
+            double_bond_multiplier,
+            triple_bond_multiplier,
+        ]
+        frame_copy = copy.deepcopy(self)
+        connectivity_matrix = np.zeros((len(self), len(self)), dtype=int)
+        distance_matrix = self.dist_matrix()
+        np.fill_diagonal(distance_matrix, np.inf)
+        for cutoff in cutoffs:
+            cutoffs = np.array(natural_cutoffs(frame_copy, mult=cutoff))
+            cutoffs = cutoffs[:, None] + cutoffs[None, :]
+            connectivity_matrix[distance_matrix <= cutoffs] += 1
+        self.connectivity = nx.from_numpy_array(connectivity_matrix)
+
+    def dist_matrix(self):
+        """
+        Experimental function, currently not in use
+        """
+        matrix = np.zeros((len(self), len(self)))
+        for i in range(1, len(self)):
+            for j in range(i, len(self)):
+                matrix[i, j] = np.linalg.norm(self.positions[i] - self.positions[j])
+        return matrix
+
+    def get_bonds(self):
+        bonds = []
+        for edge in self.connectivity.edges:
+            bonds.append((edge[0], edge[1], self.connectivity.edges[edge]["weight"]))
+        return bonds
+
+    def __len__(self):
+        if isinstance(self.numbers, np.ndarray):
+            return self.numbers.size
+        elif isinstance(self.numbers, list):
+            return len(self.numbers)
+        elif isinstance(self.numbers, int):
+            return 1
+
+    def frame_to_json(self):
```

Comment:
return type hint `-> dict`

Reply:
Also the class itself is always a `Frame` so we can rename it `to_json()` and `from_json()`. But the function is not yielding a string so `to_dict` and `from_dict` would be better.